unc-csxl/csxl.unc.edu
# Event Pagination

This PR implements pagination for the events tab. The page defaults to show upcoming events for the next month, with buttons for selecting previous or next months. You can also search all events, as long as your search is > 2 characters and searches are also limited through debouncing.

## Major Changes
### Database and Model Changes
- Added the `EventPaginationParams` model, which includes pagination parameters for specifying a start and end date, ascending/descending, order by, and filter
- Added two more events as demo data for testing this feature
### API Changes
Exposes the following API to the Angular Frontend via FastAPI:
- GET `api/events/paginate`: returns a paginated list of events based on the specified pagination parameters
### Service Changes
- Added the `list()` function to the backend `EventService` to query and paginate events based on the parameters passed in from the API call
### Frontend
- Added corresponding frontend model and service function
- Updated EventPageComponent to include the buttons for navigating between the previous and next months of events
- Updated search bar on this page to include debouncing and send search requests through API calls rather than filtering on the frontend
- Remove eventFilterPipe now that filtering is done on the backend

Picture:
![image](https://github.com/unc-csxl/csxl.unc.edu/assets/31745478/bc63ddf6-2848-4734-b37b-d177786cfd99)

## Testing
Backend: Added a new PyTest to test the list() function. Confirm 100% passing by running `pytest`
Frontend:
- Run `honcho start` and `python3 -m backend.script.reset_demo` 
- Navigate to the Events tab and see that the default dates are set to `Today - One Month from Today`. There should be four events listed in ascending order
- Select Previous Month, and date range at the top should change to last month. There should be two events listed in descending order
- Search for 'ca.' Nothing should happen. Now search for 'cads'. Four events should appear. Clearing the search bar should reset data to the correct date paginated view.

## Future Considerations
- Put pagination stuff into a widget for reusability, if that is possible. Wasn't sure how to to neatly and properly do that here since there were interactions with routes and other things.
- Note: I am no longer using the eventResolver; I was not sure if I should delete it or change my code to incorporate it back in
